File: app/src/main/java/com/system/m4/views/home/MainPageAdapter.java
package com.system.m4.views.home;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.util.SparseArray;
import android.view.ViewGroup;

/**
 * Created by eferraz on 16/07/17.
 */

class MainPageAdapter extends FragmentStatePagerAdapter {

    public static final int PAGE_MIDDLE = 6;

    private SparseArray<Fragment> registeredFragments = new SparseArray<>();

    MainPageAdapter(FragmentManager fm) {
        super(fm);
    }

    @Override
    public Fragment getItem(int position) {
        Bundle args = new Bundle();
        args.putInt(HomeFragment.RELATIVE_POSITION, position - PAGE_MIDDLE);
        args.putInt(HomeFragment.ITEM_VIEW, HomeVisibility.ALL);

        HomeFragment fragment = new HomeFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public int getCount() {
        return (PAGE_MIDDLE * 2);
    }

    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        Fragment fragment = (Fragment) super.instantiateItem(container, position);
        registeredFragments.put(position, fragment);
        return fragment;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        registeredFragments.remove(position);
        super.destroyItem(container, position, object);
    }

    @Override
    public int getItemPosition(Object object) {
        if (object instanceof HomeFragment) {
            ((HomeFragment) object).requestListTransaction();
        }
        return super.getItemPosition(object);
    }

    public Fragment getRegisteredFragment(int position) {
        return registeredFragments.get(position);
    }
}

File: app/src/main/java/com/system/m4/kotlin/view/adapters/MainPageAdapter.kt
package com.system.m4.kotlin.view.adapters

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.util.SparseArray
import android.view.ViewGroup
import com.system.m4.kotlin.view.fragments.HomeFragment

/**
 * Created by eferraz on 16/07/17.
 */

internal class MainPageAdapter(fm: FragmentManager) : FragmentStatePagerAdapter(fm) {

    private val registeredFragments = SparseArray<Fragment>()

    override fun getItem(position: Int): Fragment {
        val args = Bundle()
        args.putInt(HomeFragment.RELATIVE_POSITION, position - PAGE_MIDDLE)

        val fragment = HomeFragment()
        fragment.arguments = args
        return fragment
    }

    override fun getCount(): Int {
        return PAGE_MIDDLE * 2
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val fragment = super.instantiateItem(container, position) as Fragment
        registeredFragments.put(position, fragment)
        return fragment
    }

    override fun destroyItem(container: ViewGroup?, position: Int, `object`: Any) {
        registeredFragments.remove(position)
        super.destroyItem(container, position, `object`)
    }

    override fun getItemPosition(`object`: Any?): Int {
        if (`object` is HomeFragment) {
            `object`.requestListTransaction()
        }
        return super.getItemPosition(`object`)
    }

    fun getRegisteredFragment(position: Int): Fragment {
        return registeredFragments.get(position)
    }

    companion object {

        val PAGE_MIDDLE = 6
    }
}
